lib/spl/src/spl_draw.c

```diff
@@ -0,0 +1,953 @@
+typedef void (*DrawPlaneFunc)(fx16 s, fx16 t, fx16 offsetX, fx16 offsetY);
+
+void SPLDraw_Billboard(SPLManager *mgr, SPLParticle *ptcl);
+void SPLDraw_Child_Billboard(SPLManager *mgr, SPLParticle *ptcl);
+void SPLDraw_DirectionalBillboard(SPLManager *mgr, SPLParticle *ptcl);
+void SPLDraw_Child_DirectionalBillboard(SPLManager *mgr, SPLParticle *ptcl);
+void SPLDraw_Polygon(SPLManager *mgr, SPLParticle *ptcl);
+void SPLDraw_Child_Polygon(SPLManager *mgr, SPLParticle *ptcl);
+void SPLDraw_DirectionalPolygon(SPLManager *mgr, SPLParticle *ptcl);
+void SPLDraw_Child_DirectionalPolygon(SPLManager *mgr, SPLParticle *ptcl);
+static void SPLUtil_DrawXYPlane(fx16 s, fx16 t, fx16 offsetX, fx16 offsetY);
+static void SPLUtil_DrawXZPlane(fx16 s, fx16 t, fx16 offsetX, fx16 offsetZ);
+static void SPLUtil_RotateXYZ(fx32 sin, fx32 cos, MtxFx43 *mat);
+static void SPLUtil_RotateY(fx32 sin, fx32 cos, MtxFx43 *mat);
+
+static DrawPlaneFunc sPlaneDrawingFunctions[2] = {
+    SPLUtil_DrawXYPlane,
+    SPLUtil_DrawXZPlane,
+};
+
+static RotFunc sRotationFunctions[2] = {
+    SPLUtil_RotateY,
+    SPLUtil_RotateXYZ,
+};
+
+// Returns `TRUE` if the particle should not be drawn
+static inline BOOL SPLDraw_Setup(const SPLParticle *ptcl, const SPLManager *mgr)
+{
+    fx32 alpha = (fx32)(ptcl->visibility.baseAlpha * (ptcl->visibility.animAlpha + 1)) >> 5;
+
+    G3_PolygonAttr(
+        GX_LIGHTMASK_NONE,
+        GX_POLYGONMODE_MODULATE,
+        GX_CULL_NONE,
+        ptcl->visibility.currentPolygonID,
+        alpha,
+        mgr->miscPolygonAttr);
+
+    reg_G3_POLYGON_ATTR; // No idea why but this is needed to match
+
+    return alpha == 0;
+}
+
+static void SPLUtil_DrawXYPlane(fx16 s, fx16 t, fx16 x, fx16 y)
+{
+    G3_Begin(GX_BEGIN_QUADS);
+
+    G3_TexCoord(0, 0);
+    G3_Vtx10(x - FX32_ONE, y + FX32_ONE, 0);
+
+    G3_TexCoord(s, 0);
+    G3_Vtx10(x + FX32_ONE, y + FX32_ONE, 0);
+
+    G3_TexCoord(s, t);
+    G3_Vtx10(x + FX32_ONE, y - FX32_ONE, 0);
+
+    G3_TexCoord(0, t);
+    G3_Vtx10(x - FX32_ONE, y - FX32_ONE, 0);
+
+    G3_End();
+}
+
+static void SPLUtil_DrawXZPlane(fx16 s, fx16 t, fx16 x, fx16 z)
+{
+    G3_Begin(GX_BEGIN_QUADS);
+
+    G3_TexCoord(0, 0);
+    G3_Vtx10(x - FX32_ONE, 0, z + FX32_ONE);
+
+    G3_TexCoord(s, 0);
+    G3_Vtx10(x + FX32_ONE, 0, z + FX32_ONE);
+
+    G3_TexCoord(s, t);
+    G3_Vtx10(x + FX32_ONE, 0, z - FX32_ONE);
+
+    G3_TexCoord(0, t);
+    G3_Vtx10(x - FX32_ONE, 0, z - FX32_ONE);
+
+    G3_End();
+}
+
+static void SPLUtil_RotateXYZ(fx32 sin, fx32 cos, MtxFx43 *mat)
+{
+    fx32 C, Sm, Sp;
+
+    C = FX_MUL(FX32_ONE - cos, FX32_CONST(1 / 3.0));
+    Sm = C + FX_MUL(sin, FX32_SQRT1_3);
+    Sp = C - FX_MUL(sin, FX32_SQRT1_3);
+    C += cos;
+
+    mat->m[0][0] = C;
+    mat->m[1][0] = Sm;
+    mat->m[2][0] = Sp;
+    mat->m[3][0] = 0;
+    mat->m[0][1] = Sp;
+    mat->m[1][1] = C;
+    mat->m[2][1] = Sm;
+    mat->m[3][1] = 0;
+    mat->m[0][2] = Sm;
+    mat->m[1][2] = Sp;
+    mat->m[2][2] = C;
+    mat->m[3][2] = 0;
```

Comment:
**question:** Does it match to roll this into a double-loop or to do an explicit array-set?

Reply:
How would that work?

Reply:
Reading the code is hard, ignore me on this one